### ISSUE TITLE
Text rendering issues with multithreaded rendering

### ISSUE THREAD

**Issue by drewc5131** [panda3d/panda3d#1070]
## Description
When using the multithreaded rendering, there are two major issues with text rendering i've encountered
###  Text sometimes takes more than a frame to generate (is what I assume is happening), so it doesn't show up at first, this is easily noticeable in the FlickeringText example provided below.  
I can't put a screenshot obviously, and if you are having trouble seeing it, try increasing the number in the range() function to lower the fps down further _(globalClock set framerate to a low number causes the issue to disappear, likely because the computer is easily keeping up far above that lower number, so spawning objects puts more load)_

### Another issue is that editing text with an input dialog is very wonky

![image](https://user-images.githubusercontent.com/31282527/101950145-c7fef080-3bc2-11eb-8612-f3bf993e44f3.png)

To reproduce this in the InputIssue example i provided, enter text, and then start to backspace it. You should see overlapping text
It seems to only happen under these circumstances in my testing:
* I need to slow the framerate down to a realistic framerate, so having an empty scene rendering at 2000 fps makes this issue not occur, so like the FlickeringText example, i spawn 100 instances of a model to drop the fps down to a couple hundred
* I need to modify the text through the app at some point, so in this example, everytime the text is edited, i alternate between being normal black text, and slanted red text, based on the length of the string inputted. 

### I have provided a zip file that includes two test python files for both issues, the zip also includes a model with a texture to help slow down the fps of the engine as I stated was required earlier
[TextRenderingIssues.zip](https://github.com/panda3d/panda3d/files/5681427/TextRenderingIssues.zip)

## Environment
While this has occurred in every version I've ever tested multithreading with, and it occurs on built clients for games on various platforms on various versions of panda3d / python, I will list my current env

 * Operating system: Windows 10 1909
 * System architecture: x86-64
 * Panda3D version: 1.11.0
 * Installation method: Built from source
 * Python version: 3.9

**Comment by serkkz** [comment]
You don't need to slow down the game using the model. You can do this in the configuration file.

```
clock-mode limited
clock-frame-rate 20
```

**Comment by serkkz** [comment]
A minimal example that shows the problem.
```py
from panda3d.core import loadPrcFileData

loadPrcFileData('', 'threading-model Cull/Draw')
#loadPrcFileData('', 'clock-mode limited')
#loadPrcFileData('', 'clock-frame-rate 87')

from direct.showbase.ShowBase import ShowBase
from direct.gui.DirectGui import OnscreenText

class Test (ShowBase):
    def __init__(self):

        ShowBase.__init__(self)

        self.frameRateMeter = OnscreenText()
        
        taskMgr.add(self.updateFrameRateMeter, 'test')

        for i in range(100):
            house = loader.loadModel('panda')
            house.reparentTo(render)

    def updateFrameRateMeter(self, task):
        self.frameRateMeter.setText(str(globalClock.getDt()))
        return task.cont

test = Test()
test.run()
```
And this is not related to the frame rate.

**Comment by rdb** [comment]
Thanks for the test cases.

In @serkkz's test case, I think the problem is that the text doesn't appear until one frame *after* the text has been generated.  So if the text gets updated every frame, it won't appear.  This is a clear bug.

**Comment by serkkz** [comment]
It depends on how many models overlap.

```py
from panda3d.core import loadPrcFileData

loadPrcFileData('', 'threading-model Cull/Draw')
#loadPrcFileData('', 'clock-mode limited')
#loadPrcFileData('', 'clock-frame-rate 87')

from direct.showbase.ShowBase import ShowBase
from direct.gui.DirectGui import OnscreenText

class Test (ShowBase):
    def __init__(self):

        ShowBase.__init__(self)

        self.text = OnscreenText()
        
        taskMgr.add(self.update_text, 'test')

        for i in range(100):
            model = loader.loadModel('panda')
            model.reparentTo(render)
            model.setPos(i*40, 0, 0)

    def update_text(self, task):
        self.text.setText(str(globalClock.getDt()))
        return task.cont

test = Test()
test.run()
```
Now you can use the camera rotation to track the problem.

**Comment by rdb** [comment]
I suspected a race condition, and interestingly it appears that TextNode.setText and friends are not thread-safe, because they are part of the base TextEncoder interface.

Interestingly, though, putting a mutex around TextNode.setWtext doesn't seem to fix it.  Hmm.

**Comment by rdb** [comment]
Actually, none of the TextNode members are currently pipeline-cycled.  As the cull thread is busy trying to prepare the previous frame for rendering, the main thread is already setting the text for the next frame.  This can lead to situations where text from the wrong frame is being rendered.

This isn't too hard to fix, but the fix can't go into 1.10.8, it'll have to be fixed on the master branch.

You can work around this by using multiple TextNode objects and switching between them whenever you change the text.

**Comment by drewc5131** [comment]
> 
> 
> You don't need to slow down the game using the model. You can do this in the configuration file.
> 
> ```
> clock-mode limited
> clock-frame-rate 20
> ```

did you even read any of my post? i quite clearly stated that it did not occur when using that method.

**Comment by rdb** [comment]
This is a workaround that should work until the TextNode issue is fixed:
```python
    def updateFrameRateMeter(self, task):
        self.frameRateMeter.removeNode()
        self.frameRateMeter = OnscreenText(mayChange=False)
        self.frameRateMeter.setText(str(task.time))
        return task.cont
```

**Comment by serkkz** [comment]
@drewc5131 I don't read your texts, they usually don't contain the essence. In your post, only the code was valuable.

**Comment by rdb** [comment]
Making TextNode pipeline-cycled did indeed solve the TextNode issues.  I will be checking that in shortly.

However, the DirectEntry issues are different.  I will look into fixing these too.  There is a hack you can do in the meantime to work around the problem; the following code will force the text to be generated on the App thread:
```python
        taskMgr.add(self.updateEntry)

    def updateEntry(self, task):
        self.inputDial.guiItem.setNumLines(1)
        return task.cont
```

**Comment by rdb** [comment]
I added pipeline-cycling to TextNode, which fixes this issue.  Note that changes to some text properties are still not yet cycled (so these may not be properly synchronized between the stages), but this can be worked around by using embedded text properties.  The changes are in the master branch.

The PGEntry fixes are more complicated.  I have no time to do a proper fix now, but I did check in a workaround, which is to force the text to always be generated on the App thread right after the text is modified (if pipelining is enabled).  This workaround *can* go in 1.10.8, because it doesn't change the class layout.

Let me know (preferably with an updated test case) if you still encounter any issues.

**Comment by drewc5131** [comment]
Issue 1 is resolved, issue 2 still occurs, same test case, no changes have been made
![image](https://user-images.githubusercontent.com/31282527/102001150-9ff3b800-3cbc-11eb-851e-86171042648a.png)

**Comment by drewc5131** [comment]
I'm now noticing an issue where the frame rate meter is showing up as 0.0 fps most of the time. It shows up for like a frame as the proper framerate, but then the very next frame will show 0.0. I've set the meter update interval to `frame-rate-meter-update-interval 0.1` for faster and more accurate reporting

**Comment by drewc5131** [comment]
And I meant to confirm this a while ago, but yes the PGEntry issues seem to be gone with that fix :)

**Comment by rdb** [comment]
Yup, you're right, there's an issue with `set_text()` when called from the cull stage (as FrameRateMeter does).  Fixed in 8dd3301c43669cff5f2ede41e5ff774cc1303eb4.
